base/utf.jl

```diff
@@ -0,0 +1,880 @@
+                @get_continuation!(ch, byt, str, pos)
+                @get_continuation!(ch, byt, str, pos)
+                # check for surrogate pairs, make sure correct
+                if is_surrogate_char(ch)
+                    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, pos-2, ch)
+                    # next character *must* be a trailing surrogate character
+                    (pos + 3 > len) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos-2, ch)
+                    byt = str[pos += 1] ; (byt != 0xed) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, byt)
+                    surr::UInt32 = 0xd
+                    @get_continuation!(surr, byt, str, pos)
+                    @get_continuation!(surr, byt, str, pos)
+                    !is_surrogate_trail(surr) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos-2, surr)
+                    (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos-2, surr)
+                    flags |= UTF_SURROGATE
+                    cnt4 += 1
+                elseif ch > 0x07ff
+                    cnt3 += 1
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    flags |= UTF_LONG
+                    cnt2 += 1
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            elseif ch < 0xf5
+                # 4-byte UTF-8 sequence (i.e. characters > 0xffff)
+                (pos + 3 > len) && utf_errfunc(UTF_ERR_SHORT, pos, ch)
+                ch &= 0x07
+                @get_continuation!(ch, byt, str, pos)
+                @get_continuation!(ch, byt, str, pos)
+                @get_continuation!(ch, byt, str, pos)
+                if ch > 0x10ffff
+                    utf_errfunc(UTF_ERR_INVALID, pos-3, ch)
+                elseif ch > 0xffff
+                    cnt4 += 1
+                elseif is_surrogate_char(ch)
+                    utf_errfunc(UTF_ERR_SURROGATE, pos-3, ch)
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    # This is an overly long encode character
+                    flags |= UTF_LONG
+                    if ch > 0x7ff
+                        cnt3 += 1
+                    elseif ch > 0x7f
+                        cnt2 += 1
+                    end
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            else
+                utf_errfunc(UTF_ERR_INVALID, pos, ch)
+            end
+        end
+    end
+    return UTF_String_Counts(cntT, cnt2, cnt3, cnt4, flags | (cnt3 == 0 ? 0 : UTF_UNICODE3) | (cnt4 == 0 ? 0 : UTF_UNICODE4))
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in a UTF-16 string
+
+@param[in]  str     Vector{UInt16}
+@param[in]  options flags to determine error handling (default 0)
+
+@return     (total characters, 2-byte, 3-byte, 4-byte, flags)
+@throws     ArgumentError
+""" ->
+=#
+function check_string_utf16(str::Vector{UInt16}, len::Int)
+    local ch::UInt32
```

Comment:
You generally don't declare variables as `local` in Julia. And specifying the type is avoided as much as possible (also elsewhere in the file).